Use Lolex to implement fake timers

SimenB opened the pull request:
<!-- Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. The two fields below are mandatory. -->

<!-- Please remember to update CHANGELOG.md in the root of the project if you have not done so. -->

**Summary**
Fixes #5165.

This PR switches out Jest's custom fake timers implementation with Lolex. This greatly simplifies things on our side (as we just need to proxy their API through the Jest object), while bringing many features:

- Faking of `Date`
- Faking of `requestAnimationFrame` and soon `requestIdleCallback` (https://github.com/sinonjs/lolex/issues/218)
- Faking of `performance.now`
- `queueMicrotask`: https://github.com/sinonjs/lolex/pull/217

Note that this might mean that previously working tests now will fail because more timers are faked. We can consider allowing `jest.useFakeTimers` to get passed an array matching Lolex's `toFake` if people want to specify what to mock and what not to. See https://github.com/sinonjs/lolex/blob/master/Readme.md#var-clock--lolexinstallconfig

It brings some behavioral (aka breaking) changes as well:

- `runAllPendingTimers` in Jest currently do _not_ trigger new timers scheduled while running to the last scheduled one. Lolex does (which matches what would happen in real code). Fixes #3465
- `runAllImmediates` is removed. There is no way that real code could trigger an immediate without triggering other timers. Users can approximate this by doing `jest.advanceTimersByTime(0)`, but that will also trigger timers.

Other breaking changes:
- Callbacks scheduled using `setImmediate` or `process.nextTick` was scheduled using the real APIs in addition to the fake one, and Jest made sure that they only fired once (by keeping track). This has been removed - if you're faking time, then you're faking time
- Globals that were scheduled for removal since Jest was open sourced in 2014 are removed (this landed in #7285)
- Jest added spies to the fake timers automatically. We no longer do that, people need to do e.g. `jest.spyOn(global, 'setTimeout')` themselves.
- Timers scheduled to trigger after `0` ms will be coerced to `1`, see https://github.com/sinonjs/lolex/issues/166
- `runWithRealTimers` removed - it was undocumented and unused at FB

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->

**Test plan**
Green CI. No new tests are added as they'd be testing Lolex and not our own code. I've removed old tests that either didn't make sense or were testing features we've removed.

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes UI. -->

SimenB commented on flow-typed/npm/lolex_v2.x.x.js:
```diff
@@ -0,0 +1,42 @@
+// flow-typed signature: dca3d1bef4283f0961b10b93dd685656
```
quickly thrown together, should probably upstream it to flow-typed

SimenB replied:
someone did it :D

https://github.com/flow-typed/flow-typed/pull/2555

thymikee commented:
I think we can start with not-purging FakeTimers and a temporary new environment and break it some time later.

cpojer commented:
I checked, we don't use `runWithRealTimers`.

SimenB commented:
Cool, so sinonjs/lolex#146 and sinonjs/lolex#147 should be enough for us, then? That's awesome

cpojer commented:
We can try running this PR, once it's fully iterated, on the FB codebase to uncover more issues.

Chudesnov commented:
> runAllImmediates()
> is probably an anti-pattern as it breaks the abstraction.

But it is also the only way to advance code based on Promises in tests (scheduled with setImmediate, [as mentioned](https://github.com/facebook/jest/issues/5165#issuecomment-353784870) in #5165 by @cpojer) without triggering other timers.

cpojer commented:
Are we planning to ship this or shall we close this PR for now?

SimenB commented:
There has been some movement in Lolex, I need to test if it's enough. Will look into it this weekend 🙂

SimenB commented:
I've rebased this and added the two potential changes in Lolex to help support our use case.

From my testing now, it's not enough, so I reported back on one of the PRs over there (https://github.com/sinonjs/lolex/pull/157).

SimenB commented:
I've rebased now, and tested with https://github.com/sinonjs/lolex/pull/164 merged. It's pretty close (I sent https://github.com/sinonjs/lolex/pull/165 for a quick fix of one issue found) which works really well, but I'm having a few issues:

1. `runOnlyPendingTimers` (which I've implemented as calling Lolex' `runToLast`) runs all new timers added, not just the previously pending ones. Is that ok? I suppose this was known, but it's pretty breaking (I mentioned this in the OP as well, but it's been a while)
1. The `throws before allowing infinite recursion` test doesn't throw, but I think that's just a matter of fixing the test - the functionality works normally
1. `runAllTicks` still doesn't work... Calling Lolex' new `runMicrotasks` just makes the test hang forever, and I'm not sure why

SimenB commented:
Another key difference is that the timer function are no longer jest mocks when faked - people will have to do `jest.spyOn` themselves. Is that OK? Or should we spy on the functions by default?

SimenB commented:
Figured out issue nr 3 above, it's the infinites recursion not making it bail. Might be same issue as number 2

SimenB commented:
Dug into number 2 above, and I think the thing there is that the recursion check is only for `runAll`, not `tick`. And Lolex has code forcing timers added during `tick` to have a timeout of `1`, meaning when we advance by `50` ms, we get 51 timeouts. Not sure if it's by design, opening up an issue

thymikee commented:
Regarding this comment https://github.com/sinonjs/lolex/issues/166#issuecomment-387310192, is it web only or Node acts in the same way?

SimenB commented:
My PRs to Lolex has been merged and released, so I updated this PR. Getting quite close now 🙂

SimenB commented:
@cpojer would it be possible to get an example of how you use fake timers to drive promises at FB? Would be great to add it as an integration test or something

cpojer commented:
cc @mjesun @rubennorte who can give more comments and/or run this diff on FB infra.

benjamingr commented:
> I think I'll implement setImmediate manually, in order for us to keep the capability of driving promises with them without touching timers. 

Well, `setImmediate` works differently between IE and Node in terms of semantics - we can implement it as a `setTimeout(fn, 0)` in lolex pretty easily and it would _roughly_ approximate the semantics in both.

Pop quiz, what is the output of the following code? 

```js
setImmediate(() => setImmediate(() => console.log(2)));
setTimeout(() => console.log(1), 0)
```

Anyway - no strong feelings from me about lolex supporting setImmediate - I typically use lolex to test timeouts of things like "cancel this action after 2000ms have passed if some condition is not met" so I rarely need `setImmediate`.

----

> Other than that, I think the only thing we have to land is what to do about runAllPendingTimers. Jest currently only run all scheduled timers, ignoring new ones added during the run. Lolex (with its runToLast function) figures out how many ticks it has to run to run all timers, and ticks that far, triggering all timers added during the way there. I think that's a way more sane approach (as it simulates what would happen in real code), but it's pretty breaking.

To be honest that just sounds like a rather serious bug in Jest's fake timers, if I do:

```js
setTimeout(() => {
  console.log(1);
  setTimeout(() => console.log(2), 10);
}, 10);
setTimeout(() => console.log(3), 100);
```

Do you mean that I get 1 and 3 logged but not 2 and when calling it again I'd get 2? Doesn't that seriously break execution order in tests?

SimenB commented:
>Anyway - no strong feelings from me about lolex supporting setImmediate 

Lolex already supports `setImmediate` - the issue here is that we want to trigger them without touching any timers or ticks at all. I think it breaks the abstraction, so it's sorta a special case. But I'd need to get en example from FB's tests showing how it's used before drawing any conclusions, I think.

>Do you mean that I get 1 and 3 logged but not 2 and when calling it again I'd get 2?

Correct, see https://repl.it/@SimenB/LividNutritiousBlog for a live example. Docs: https://facebook.github.io/jest/docs/en/jest-object.html#jestrunonlypendingtimers

>If any of the currently pending macro-tasks schedule new macro-tasks, those new tasks will not be executed by this call

I agree this behavior is pretty weird, but I have no doubt thousands of tests depend on it.

The again, reading the reasoning behind the behavior

>In these scenarios, it's useful to be able to run forward in time by a single step at a time.

It seems to me like `runToLast` would handle that use case as well. Maybe it's ok as a breaking change? 🙂

SimenB commented:
@karl sorry, missed your comment! It's currently a bit stagnated due to lack of time and understanding of all requirements from FB's point of view. I honestly don't remember concretely. I wonder if #6876 helps with promises enough that it unblocks this PR.

I'll rebase, see what CI says, then try to give this another shot :)

SimenB commented:
I've rebased, and CI just aborts :( Help appreciated, I don't know when I'll have the time to dig into this (won't be for a couple of weeks at least)

mjesun commented:
cc @rubennorte, we just talked about that. Mind taking it from here? Most of the internal issues were fixed already.

SimenB commented:
Thanks!

I can reproduce the abort locally: `yarn jest packages/jest-util/src/__tests__/fake_timers.test.js`. Interestingly, it just hangs forever if not running in band - my guess is that the worker is restarted.

Now, I don't know _why_ it aborts, I don't think it used to do that :(

SimenB commented:
Found the bug: https://github.com/sinonjs/lolex/pull/208

Triggered here because we do `now: Date.now()` when installing the timers

thejohnfreeman commented:
Would it be possible to address [this issue](https://github.com/sinonjs/lolex/issues/213), but for Jest, in this PR?

SimenB commented:
It's currently in Jest as a global. IIRC FB needed it though, so we should expose a real API for it.

https://github.com/facebook/jest/blob/d59a4d33b6aff50df3f2a307d71797a5d0e5f298/packages/jest-util/src/fake_timers.js#L126

SimenB commented:
Might as well post an update. We're gearing up for a new major, and I would _really_ love to land this as part of that.

There are 2 main blockers (beyond behaviour change for `runOnlyPendingTimers`, which I think is a fine change - current behavior is plain wrong)

1. By default we spy on `setTimeout` et.al. when mocking them, so people out of the box can do `expect(setTimeout).toHaveBeenCalled...`. This works fine, up until the point when somebody calls `jest.restoreAllMocks`, and the spies are cleared out. I think we either have to remove auto-spying or add an API to `jest-mock` where you can subscribe to `restore` calls so we can re-add the spies. See my last 2 commits (made some weeks apart) with the second one reverting the first one since I got confused.
2. The removal of `runAllImmediates`. I've mentioned this above, so I won't go into details, but I think it's a bad API and should be removed. Real code would never trigger immediates without also triggering microticks, so it breaks the abstraction. However, FB uses it to drive fake promises (https://github.com/facebook/jest/pull/7071#issuecomment-426205082), so I'm not sure if it's feasible to remove it. See earlier comments for more discussion.

This PR should also upgrade to Lolex 3, the breaking change (disallow negative ticks) doesn't impact Jest at all

rubennorte commented:
> The removal of runAllImmediates. I've mentioned this above, so I won't go into details, but I think it's a bad API and should be removed. Real code would never trigger immediates without also triggering microticks, so it breaks the abstraction. However, FB uses it to drive fake promises (#7071 (comment)), so I'm not sure if it's feasible to remove it. See earlier comments for more discussion.

We do use `setImmediate` but that doesn't mean we can't live without `runAllImmediates`. We're not using it so frequently and might be able to replace them with `runAllTimers` or adapt the tests.

SimenB commented:
Aren't you busy watching react conf? 😀 

Anyways, that's awesome! To celebrate, I'll rip it out then, as it's part of the failing tests 🙂 I'll also update the tests for `runPendingTimers`.
After that, we'll just have to figure out how to attach spies to timer functions. I'm currently leaning towards not spying on them by default, forcing people - if they care - to add a spy themselves.

SimenB commented:
Dug into the failing tests a bit now, and they're asserting what I would call really weird behavior.

See e.g. the current fake tick in master:

https://github.com/facebook/jest/blob/4941a974e303db63e6ee5832f3737e69a138de18/packages/jest-util/src/fake_timers.js#L399-L424

It schedules a _real_ `process.nextTick` in addition to the fake one, making sure to only trigger one, if the other hasn't done so already. I don't understand that behavior - either you want to fake time, or you don't want to. Faking timers, but allowing them to fire on their own seems super weird to me.

There are tests asserting on this behavior:
https://github.com/facebook/jest/blob/4941a974e303db63e6ee5832f3737e69a138de18/packages/jest-util/src/__tests__/fake_timers.test.js#L159-L274

I've just removed those tests from this branch, hopefully we have consensus that that's the correct solution. I'm afraid it'll break a billion tests, though, because people are relying on kinda faking time, but not really...

SimenB commented on packages/jest-util/src/__tests__/fake_timers.test.js:
```diff
@@ -665,137 +561,6 @@ describe('FakeTimers', () => {
     });
   });
 
-  describe('runWithRealTimers', () => {
```
FB doesn't use this, so removed

SimenB commented:
OK, this is finally ready for review. I've updated the docs, and left some inline comments.

I think we should setup a table in the docs somewhere showing which functions in Jest corresponds to which  functions in Lolex. While we've kept our function names here, we might want to delegate some of the documentation responsibility to Lolex (who's currently revamping their docs as well: https://github.com/sinonjs/lolex/issues/219)

When we get to writing the blog post for Jest 24, I'll make sure to add a paragraph or 2 about the benefits and tradeoffs for switching to Lolex.

SimenB commented on packages/jest-util/src/fake_timers.js:
```diff
@@ -123,423 +46,96 @@ export default class FakeTimers<TimerRef> {
     global.mockAdvanceTimersByTime = this.advanceTimersByTime.bind(this);
```
https://github.com/facebook/jest/issues/5165#issuecomment-353784870

>FakeTimers also exposes some legacy globals which I’d be happy to kill completely and codemod internally at FB.

thymikee replied:
Please do so, they're easily replaceable

SimenB replied:
Done!

thymikee commented on docs/JestObjectAPI.md:
```diff
@@ -364,6 +360,16 @@ Executes only the macro-tasks that are currently pending (i.e., only the tasks t
 
+### `.jest.setSystemTime()`
+
+Set the current system time used by fake timers, affecting the time returned by `ned Date` et.al.
```
typo: `new Date()`

thymikee commented on docs/JestObjectAPI.md:
```diff
@@ -364,6 +360,16 @@ Executes only the macro-tasks that are currently pending (i.e., only the tasks t
+
+### `.jest.getRealSystemTime()`
+
+When mocking time, `Date.now()` will also be mocked. If you for some reason needs access to the real current time, you can invoke this function.
```
typo: needs -> need

thymikee commented on packages/jest-util/src/__tests__/fake_timers.test.js:
```diff
@@ -632,26 +491,45 @@ describe('FakeTimers', () => {
         'mock3',
-
+        'mock1',
+        'mock2',
```
Can we extract this (509:518) to `runOrderFirst` and below to `runOrderSecond`? I think we could make it easier to read, as above and lines 495:504 are the same part of `runOrder` queue.

SimenB replied:
Went with


```js
expect(runOrder).toEqual([
  ...firsRunOrder,
  'mock2',
  'mock1',
  'mock2',
  'mock2',
  'mock3',
  'mock5',
  'mock1',
  'mock2',
]);
```

thoughts?

thymikee commented on docs/JestObjectAPI.md:
```diff
@@ -364,6 +360,16 @@ Executes only the macro-tasks that are currently pending (i.e., only the tasks t
+
+Set the current system time used by fake timers, affecting the time returned by `ned Date` et.al.
+
+See the [Lolex docs](https://github.com/sinonjs/lolex/blob/master/Readme.md#clocksetsystemtimenow) for more information.
```
I think we could use the wording from the current Lolex docs instead of (or next to) linking to it.

SimenB replied:
Shortened our part, kept the link. thoughts?

thymikee replied:
I think I'd rather like:

>Set the current system time used by fake timers. Simulates a user changing the system clock while your program is running. It affects the current time but it does not in itself cause e.g. timers to fire; they will fire exactly as they would have done without the call to `jest.setSystemTime()`.

Really no need to link to such a short explanation. However, somewhere at the top of Fake Timer docs, we could state that we use Lolex and here's the docs to it if you're curious, but you shouldn't :p

SimenB replied:
Went for that.

We really should group information in that document - one section for module mocks, one for functions pies, one for timers etc. Now it's all jumbled up

thymikee replied:
Agreed. @rickhanlonii ideas on making that happen?

SimenB replied:
#7278

SimenB commented on packages/jest-runtime/src/index.js:
```diff
@@ -445,20 +445,22 @@ class Runtime {
-      if (envGlobal.mockClearTimers) {
-        envGlobal.mockClearTimers();
+      if (this._environment.fakeTimers) {
+        this._environment.fakeTimers.clearAllTimers();
```
is this expected?

thymikee replied:
This behavior is there from the very first commit: https://github.com/facebook/jest/commit/88a94d5d1bc1f387317a3068bf510ab992c5dc64#diff-c57162543a4f0bc311cba66ba52baf7bR936

BTW, look what's 5 yrs deprecated too: https://github.com/facebook/jest/commit/88a94d5d1bc1f387317a3068bf510ab992c5dc64#diff-b689d4f4c53abcb13c3c1d211b4d8bc5R55 😄

SimenB replied:
haha, "very soon" indeed

rubennorte commented on packages/jest-util/src/fake_timers.js:
```diff
@@ -9,537 +9,123 @@
-    global.mockAdvanceTimersByTime = this.advanceTimersByTime.bind(this);
-    global.mockRunTimersRepeatedly = this.runAllTimers.bind(this);
-    global.mockClearTimers = this.clearAllTimers.bind(this);
-    global.mockGetTimersCount = () => Object.keys(this._timers).length;
```
You're removing `global.mockGetTimersCount` and there isn't an alternative public API for it. Can you add it as a method in the `jest` object? We still use it at Facebook.

SimenB replied:
Added `getTimersCount` in 266a3c3e66329b093cf6c87394d6dee33f366a10

rubennorte commented on packages/jest-util/src/fake_timers.js:
```diff
@@ -9,537 +9,123 @@
-    );
-    for (const uuid in this._timers) {
-      delete this._timers[uuid];
+    if (this._fakingTime) {
```
Why isn't this guarded with `checkFakeTimers` as the rest of methods?

SimenB replied:
good catch

SimenB replied:
Actually, removing it breaks stuff due to https://github.com/facebook/jest/pull/5171#discussion_r228455544

thymikee commented on docs/JestObjectAPI.md:
```diff
@@ -443,6 +443,10 @@ Removes any pending timers from the timer system.
 
+### `jest.getTimerCount()`
+
+Returns the number of fake timers still left to run
```
add a dot at the end

rubennorte commented:
> runAllPendingTimers in Jest currently do not trigger new timers scheduled while running to the last scheduled one. Lolex does (which matches what would happen in real code). Fixes #3465

Wouldn't this cause infinite recursion in some cases?

> Note that this might mean that previously working tests now will fail because more timers are faked. We can consider allowing jest.useFakeTimers to get passed an array matching Lolex's toFake if people want to specify what to mock and what not to. See https://github.com/sinonjs/lolex/blob/master/Readme.md#var-clock--lolexinstallconfig

I don't think that granularity is beneficial. Faking all time related built-ins is the right thing to do when `useFakeTimers` is used. We should provide a way to "unfake" them to make it easier for people to migrate.

> Jest added spies to the fake timers previously. We no longer do that, people need to do e.g. jest.spyOn(global, 'setTimeout') themselves.

I still have to test how this would behave in FB tests.

SimenB commented:
>Wouldn't this cause infinite recursion in some cases?

It still bails at `maxLoop` (`100000` by default)

>> Note that this might mean that previously working tests now will fail because more timers are faked. We can consider allowing jest.useFakeTimers to get passed an array matching Lolex's toFake if people want to specify what to mock and what not to. See https://github.com/sinonjs/lolex/blob/master/Readme.md#var-clock--lolexinstallconfig

>I don't think that granularity is beneficial. Faking all time related built-ins is the right thing to do when `useFakeTimers` is used.

I totally agree faking all is the right thing to do 🙂

>We should provide a way to "unfake" them to make it easier for people to migrate.

Like `jest.unfakeTimer('performance')` or something like that?

AlexLomm commented:
Hello, are there any updates on this issue?

thymikee commented:
@AlexLomm it's still being tested internally at FB. Let's be patient, there's no need to rush things ;)

AlexLomm commented:
@thymikee No rush, just being curious :). Thanks for such a quick reply and keep up the amazing work! :)

thymikee commented:
How do we stand with this one? Any feedback from internal testing? cc @rubennorte @rickhanlonii @aaronabramov

SimenB commented:
Non-breaking alternative: #7776

SimenB commented:
CI is very unhappy due to https://github.com/sinonjs/lolex/issues/232

Shahor commented:
Hi there 👋 

Any chance of seeing this hitting release stage?

SimenB commented:
This is too breaking for FB as they rely on the broken semantics of the current implementation. So the current plan is to land a Lolex implementation in Jest 26 which will be the default if fake timers are used, and make the current implementation opt-in by saying `jest.useFakeTimers('legacy')`. That's in #7776, only flipped defaults.

I don't want to land Lolex as opt-in first to avoid juggling extra options
